lib/label.ml

```diff
@@ -0,0 +1,202 @@
+  | Dir d -> Fmt.pf ppf "dir=%s" d
+  | Source_tree s -> Fmt.pf ppf "source-tree=%s" s
+  | File f -> Fmt.pf ppf "file=%s" f
+  | Part p -> Fmt.pf ppf "part=%s" p
+  | Env e -> Fmt.pf ppf "env=%s" e
+  | Skip -> Fmt.string ppf "skip"
+  | Non_det `Output -> Fmt.string ppf "non-deterministic=output"
+  | Non_det `Command -> Fmt.string ppf "non-deterministic=command"
+  | Version (op, v) ->
+    Fmt.pf ppf "version%a%a" Relation.pp op Ocaml_version.pp v
+  | Require_package p -> Fmt.pf ppf "require-package=%s" p
+  | Set (v, x) -> Fmt.pf ppf "set-%s=%s" v x
+  | Unset x -> Fmt.pf ppf "unset-%s" x
+
+let is_prefix ~prefix s =
+  let len_prefix = String.length prefix in
+  if String.length s > len_prefix then
+    String.equal (String.sub s 0 len_prefix) prefix
+  else false
+
+(* [is_prefix ~prefix s] is always checked before. *)
+let split_prefix ~prefix s =
+  let len_prefix = String.length prefix in
+  String.sub s len_prefix (String.length s - len_prefix)
+
+let doesnt_accept_value ~label ~value res =
+  match value with
+  | Some _ ->
+    Error (`Msg (Format.sprintf "Label `%s` does not allow a value." label))
+  | None -> Ok res
+
+let requires_value ~label ~value f =
+  match value with
+  | Some (op, v) -> f op v
+  | None -> Error (`Msg (Format.sprintf "Label `%s` requires a value." label))
+
+let requires_eq_value ~label ~value f =
+  match value with
+  | Some (Relation.Eq, v) -> Ok (f v)
+  | Some _ ->
+    let msg =
+      Format.sprintf
+        "Label `%s` requires assignment using the `=` operator." label
+    in
+    Error (`Msg msg)
+  | None -> Error (`Msg (Format.sprintf "Label `%s` requires a value." label))
+
+let of_string s =
+  let label, value = Relation.raw_parse s in
+  match label with
+  (* flags: labels without value *)
+  | "skip" ->
+    doesnt_accept_value ~label ~value Skip
+  | v when is_prefix ~prefix:"unset-" v ->
+    doesnt_accept_value ~label ~value (Unset (split_prefix ~prefix:"unset-" v))
+  (* labels requiring a value and any comparison operator *)
+  | "version" -> (
+      requires_value ~label ~value (fun op v ->
+          match Ocaml_version.of_string v with
+          | Ok v -> Ok (Version (op, v))
+          | Error (`Msg e) ->
+            let msg = Format.sprintf "Invalid `version` label value: %s." e in
+            Error (`Msg msg) ) )
+  (* non-deterministic accepts any value *)
+  | "non-deterministic" -> (
+      match value with
+      | Some (Relation.Eq, "output") -> Ok (Non_det `Output)
+      | Some (Relation.Eq, "command") -> Ok (Non_det `Command)
+      | Some (Relation.Eq, v) ->
+        let msg =
+          Format.sprintf
+            "%S is not a valid value for label `%s`. Valid values are <none>,\
+            \ %S and %S."
+            v label "command" "output"
+        in
+        Error (`Msg msg)
+      | Some _ ->
+        let msg =
+          Format.sprintf
+            "label `%s` requires assignment using the `=` operator" label
+        in
+        Error (`Msg msg)
```

Comment:
You could make a `requires_eq` function and implement `requires_eq_value` using `requires_eq` and `requires_value`.

Reply:
Good idea, done.